### ISSUE TITLE
How to train a network without noise and inference with noise

### ISSUE THREAD

**Issue by Jize-W** [IBM/aihwkit#273]
Hello,
The hardware aware training method needs to train with noise. If I want to train normally without noise, and implement the weights onto an analog network and infer with noise, what should I do?
Besides, I'm not very clear about the optional parameters of class `InferenceRPUConfig()`. Where should I check them?
![image](https://user-images.githubusercontent.com/75832609/123917041-eab4f300-d9b4-11eb-9e93-2b737eb5bd3b.png)

**Comment by maljoras** [comment]
Hi @Jize-W,
many thanks for the question. If you do not want to use any hardware-ware training you can just use any standard torch model and train the network and then convert it to analog as shown in [example 17](https://github.com/IBM/aihwkit/blob/master/examples/17_resnet34_imagenet_conversion_to_analog.py) using the `InferenceRPUConfig` and then program and drift the weights (see also [example 6](https://github.com/IBM/aihwkit/blob/master/examples/06_lenet5_hardware_aware.py#L318)):

```python
from torchvision.models import resnet34  # e.g. trained model from torch
from aihwkit.nn.conversion import convert_to_analog
from aihwkit.simuator.configs import InferenceRPUConfig

rpu_config =  InferenceRPUConfig()
# weight scaling to make sure that max weight is mapped to gmax
analog_model = convert_to_analog(resnet34, rpu_config, weight_scaling_omega=1.0)

# evaluate model after programming the weights and drifting to a certain time
analog_model.eval()
t_inference = 1000.  # desired time of drift in seconds after programming
analog_model.drift_analog_weights(t_inference)   
result = analog_model(data)    # this evaluates the programmed + drifted model on a given data 
```
In this case, all `Linear` and `Conv` layers will be converted to analog counter parts (conversion of LSTMs is not yet supported). However, note that currently the each analog layer only has one tile of the same size as the weight matrix, so it will not do e.g. multiple tile mappings of layers automatically.       

The above uses the standard setting of the forward pass and PCM statistical noise model for programming, 1/f noise and drift (defined in `rpu_config.noise_model`) and also uses global drift compensation if defined in `rpu_config.drift_compensation`. The settings of the forward pass can be adapted by setting `rpu_config.forward`.  Note that the inference pass will have two types of noises: one that is attached to the weight matrix (according to the noise model see [here](https://aihwkit.readthedocs.io/en/latest/pcm_inference.html)) and an trial-to-trial noisy analog Multiply-and-accumulate operation (defined in `rpu_config.forward`)

**Comment by Jize-W** [comment]
Hi @maljoras ,
Thank you for your detailed answer.
However, I am still confused with the rpu configurations (e.g. in the figure below). I don't know what variables I can set. Could you please provide a sheet that contains all the member varibles (i.e. possible configurations) of class `InferenceRPUConfig()` and their corresponding physical meanings? That will help me a lot.
![image](https://user-images.githubusercontent.com/75832609/124360164-e578d200-dc5a-11eb-94a2-cf90b1c2004f.png)

**Comment by maljoras** [comment]
Hi @Jize-W , please take a look at the api documentation, all members are explained there.

You can just search for InferenceRPUConfig in the documentation page. 

Best,
Malte

**Comment by rayjyh** [comment]
Hi @maljoras, 

>  However, note that currently each analog layer only has one analog tile of the same size as the weight matrix, so it will not do e.g. multiple tile mappings of layers automatically.

How will multiple tile mappings of layers affect the inference accuracy? Isn't only the ads/dac and device noise going to affect the accuracy?

**Comment by maljoras** [comment]
Hi @rayjyh,

Tiling the output dimension into multiple tiles would currently in the simulator indeed not affected the accuracy as output lines are considered to be independent, however, tiling the input dimension would. If input dimension is larger, the output will saturate more quickly. Moreover, when tiling the input dimension, a digital addition is needed to add the tile outputs together, which also might impact accuracy. Furthermore, we also plan to  introduce in future noise that depends on the amount of current and other features such as IR drop, which also will depend on the input dimension of the physical tile.
